torch/_inductor/memory.py

```diff
@@ -0,0 +1,543 @@
+        if buf_name in name_to_buf
+    )
+    max_memory = max(live_memory, output_memory)
+
+    # compute the amount of memory that is allocated when a node is scheduled
+    # and the amount of memory that can be freed when a node is scheduled
+    for i, node in enumerate(nodes):
+        node.mpi.index = i  # keep track of the original order
+        node.mpi.size = sum(buffer.mpi.size_alloc for buffer in node.get_outputs())
+        node.mpi.memory_to_free = 0
+        # 1. if a buffer read by this node is last used by this node
+        #    then the buffer can be freed
+        for buf in node.mpi.pred_buffers:
+            if buf.mpi.outdegree == 1:
+                node.mpi.memory_to_free += buf.mpi.size_free
+        # 2. if a buffer written by this node is used internally and
+        #    not needed afterwards, it can be freed
+        for buf in node.get_outputs():
+            if buf.mpi.outdegree == 0:
+                node.mpi.memory_to_free += buf.mpi.size_free
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    num_iters: int = 0
+    while num_iters < len(nodes) and nodes_to_schedule:
+        # select a node to schedule:
+        selected_node = min(
+            nodes_to_schedule,
+            key=lambda node: (
+                max(live_memory + node.mpi.size, max_memory),
+                node.mpi.size - node.mpi.memory_to_free,
+                node.mpi.index,
+            ),
+        )
+        nodes_to_schedule.remove(selected_node)
+        schedule.append(selected_node)
+        num_iters += 1
+
+        # update memory usage
+        live_memory += selected_node.mpi.size
+        max_memory = max(max_memory, live_memory)
+        live_memory -= selected_node.mpi.memory_to_free
+
+        # update successor nodes and nodes_to_schedule
+        for succ_node in selected_node.mpi.succ_nodes:
+            assert succ_node.mpi.indegree > 0
+            succ_node.mpi.indegree -= 1
+            if succ_node.mpi.indegree == 0:
+                nodes_to_schedule.add(succ_node)
+
+        # update predecessor nodes
+        for buf in selected_node.mpi.pred_buffers:
+            assert buf.mpi.outdegree > 0
+            buf.mpi.outdegree -= 1
+            if buf.mpi.outdegree == 1:
+                for succ_node in buf.mpi.succ_nodes:
+                    succ_node.mpi.memory_to_free += buf.mpi.size_free
+
+    if num_iters > len(nodes):
+        raise RuntimeError("Failed to schedule, while loop ran too long for lpmf")
+
+    return schedule
+
+
+def topological_sort_bfs(nodes: List[BaseSchedulerNode]) -> List[BaseSchedulerNode]:
+    """
+    A BFS topological sort that selects nodes whose dependencies are executed the
+    earliest. This follows a FIFO idea. Specifically, at every iteration, for each node
+    that is schedulable, we gather the time when its predecessors nodes are executed,
+    and this sorted list of execution times of predecessors nodes defines the priority.
```

Comment:
we might want to emphasize it's execution order and not execution time (which could be confused with length of execution):
```suggestion
    that is schedulable, we gather the order in which its predecessor nodes are executed,
    and this sorted list of execution orders of predecessor nodes defines the priority.
```